src/test-harness/SparkLiquidityLayerTests.sol

```diff
@@ -241,17 +295,181 @@ abstract contract SparkLiquidityLayerTests is SpellRunner {
+        MainnetController prevController;
+        MainnetController controller;
+        uint256[] depositAmounts;
+        uint256 minLPAmount;
+        uint256[] withdrawAmounts;
+        uint256[] rates;
+        bytes32 swapKey;
+        bytes32 depositKey;
+        bytes32 withdrawKey;
+        uint256 minAmountOut;
+        uint256 lpBalance;
+        uint256 smallerMaxSlippage;
+    }
+
+    function _testCurveOnboarding(
+        address pool,
+        uint256 expectedDepositAmountToken0,
+        uint256 expectedSwapAmountToken0,
+        uint256 maxSlippage,
+        RateLimitData memory swapLimit,
+        RateLimitData memory depositLimit,
+        RateLimitData memory withdrawLimit
+    ) internal {
+        // Avoid stack too deep
+        CurveOnboardingVars memory vars;
+        vars.pool  = ICurvePoolLike(pool);
+        vars.rates = ICurvePoolLike(pool).stored_rates();
+
+        assertEq(vars.pool.N_COINS(), 2, "Curve pool must have 2 coins");
+
+        vars.ctx            = _getSparkLiquidityLayerContext();
+        vars.controller     = MainnetController(vars.ctx.controller);
+        vars.prevController = MainnetController(vars.ctx.prevController);
+
+        vars.depositAmounts = new uint256[](2);
+        vars.depositAmounts[0] = expectedDepositAmountToken0;
+        // Derive the second amount to be balanced with the first
+        vars.depositAmounts[1] = expectedDepositAmountToken0 * vars.rates[0] / vars.rates[1];
+
+        vars.minLPAmount = (
+            vars.depositAmounts[0] * vars.rates[0] +
+            vars.depositAmounts[1] * vars.rates[1]
+        ) * maxSlippage / 1e18 / vars.pool.get_virtual_price();
+
+        vars.swapKey     = RateLimitHelpers.makeAssetKey(vars.controller.LIMIT_CURVE_SWAP(),     pool);
+        vars.depositKey  = RateLimitHelpers.makeAssetKey(vars.controller.LIMIT_CURVE_DEPOSIT(),  pool);
+        vars.withdrawKey = RateLimitHelpers.makeAssetKey(vars.controller.LIMIT_CURVE_WITHDRAW(), pool);
+
+        _assertRateLimit(vars.swapKey,     0, 0);
+        _assertRateLimit(vars.depositKey,  0, 0);
+        _assertRateLimit(vars.withdrawKey, 0, 0);
+
+        if (vars.prevController == vars.controller) {
+            // Only check if we are not doing a controller upgrade
+            assertEq(vars.prevController.maxSlippages(pool), 0);
+        }
+
+        executeAllPayloadsAndBridges();
+
+        _assertRateLimit(vars.swapKey,     swapLimit);
+        _assertRateLimit(vars.depositKey,  depositLimit);
+        _assertRateLimit(vars.withdrawKey, withdrawLimit);
+
+        assertEq(vars.controller.maxSlippages(pool), maxSlippage);
+
+        if (depositLimit.maxAmount != 0) {
+            // Deposit is enabled
+            assertGt(vars.depositAmounts[0], 0);
+            assertGt(vars.depositAmounts[1], 0);
+
+            deal(vars.pool.coins(0), address(vars.ctx.proxy), vars.depositAmounts[0]);
+            deal(vars.pool.coins(1), address(vars.ctx.proxy), vars.depositAmounts[1]);
+
```

Comment:
A bit redundant but good for visibility of before/after:

```solidity
assertEq(IERC20(vars.pool.coins(0)).balanceOf(address(vars.ctx.proxy)), vars.depositAmounts[0]);
assertEq(IERC20(vars.pool.coins(1)).balanceOf(address(vars.ctx.proxy)), vars.depositAmounts[1]);
```

Reply:
added